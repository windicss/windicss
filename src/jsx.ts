/* eslint-disable @typescript-eslint/no-empty-interface */
export type UtilityNames =
  | 'p'
  | 'm'
  | 'w'
  | 'h'
  | 'z'
  | 'border'
  | 'grid'
  | 'flex'
  | 'bg'
  | 'text'
  | 'font'
  | 'opacity'
  | 'animate'
  | 'transition'
  | 'transform'
  | 'align'
  | 'justify'
  | 'items'
  | 'block'
  | 'content'
  | 'pos'
  | 'box'
  | 'overflow'
  | 'underline'
  | 'list'
  | 'gradient'
  | 'divide'
  | 'gap'
  | 'ring'
  | 'icon'
  | 'container'
  | 'space'
  | 'table'
  | 'order'
  | 'place'
  | 'display'
  | 'shadow'
  | 'blend'
  | 'filter'
  | 'backdrop'
  | 'cursor'
  | 'outline'
  | 'select'

export type VariantNames =
  | 'hover'
  | 'active'
  | 'focus'
  | 'enabled'
  | 'dark'
  | 'light'
  | 'sm'
  | 'lg'
  | 'md'
  | 'xl'
  | 'xxl'
  | 'first'
  | 'last'
  | 'child'
  | 'root'
  | 'before'
  | 'after'
  | 'all'

export type AttributifyNames<Prefix extends string = ''> =
  | `${Prefix}${UtilityNames}`
  | `${Prefix}${VariantNames}`
  | `${Prefix}${VariantNames}:${UtilityNames}`

export interface AttributifyAttributes extends Partial<Record<AttributifyNames, string>> {}
