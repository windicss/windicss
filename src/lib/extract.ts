import { Utility } from './utilities/handler';
import { deepCopy } from '../utils/tools';
import { Style, Property } from '../utils/style';
import { pluginOrder } from '../config/order';
import { staticUtilities, dynamicUtilities } from './utilities';
import type { Processor } from './index';

export function generateStaticStyle(processor: Processor, className:string, addComment = false): Style | undefined {
  // eslint-disable-next-line no-prototype-builtins
  if (!staticUtilities.hasOwnProperty(className))
    return;

  const style = new Style('.' + className);
  const comment = addComment ? className : undefined;
  const { utility, meta } = staticUtilities[className];
  for (const [key, value] of Object.entries(utility)) {
    style.add(
      Array.isArray(value)
        ? value.map((i) => new Property(key, i, comment))
        : new Property(key, value, comment)
    );
  }
  if (processor._plugin.core && !processor._plugin.core[meta.group]) return;
  return style.updateMeta('utilities', meta.group, pluginOrder[meta.group as keyof typeof pluginOrder], meta.order, true);
}

export default function extract(
  processor: Processor,
  className: string,
  addComment = false,
  prefix?: string,
): Style | Style[] | undefined {
  if (prefix && !className.startsWith(prefix)) return;

  // handle static base utilities
  if (!prefix && className in staticUtilities) return generateStaticStyle(processor, className, addComment);
  if (prefix && className.startsWith(prefix)) {
    className = className.replace(new RegExp(`^${prefix}`), '');
    if (className in staticUtilities) return generateStaticStyle(processor, className, addComment);
  }
  // handle static plugin utilities & components
  const staticPlugins = { ...processor._plugin.utilities, ...processor._plugin.components, ...processor._plugin.shortcuts };
  if (className in staticPlugins) return deepCopy(staticPlugins[className]);

  const utility = new Utility(className, processor._handler);


  // handle dynamic plugin utilities
  for (const [key, generator] of Object.entries(processor._plugin.dynamic)) {
    if (className.match(new RegExp(`^-?${key}`))) {
      let style = generator(utility);
      if (style instanceof Property) style = style.toStyle(utility.class);
      if (style && addComment)
        Array.isArray(style)
          ? style.map((i) => i.property.forEach((p) => (p.comment = className)))
          : style.property.forEach((p) => (p.comment = className));
      if (style) return style;
    }
  }

  // handle dynamic base utilities
  const matches = className.match(/\w+/);
  const key = matches ? matches[0] : undefined;
  // eslint-disable-next-line no-prototype-builtins
  if (key && dynamicUtilities.hasOwnProperty(key)) {
    let style = dynamicUtilities[key](utility, processor.pluginUtils);
    if (!style) return;
    if (processor._plugin.core && !processor._plugin.core[Array.isArray(style) ? style[0].meta.group : style.meta.group]) return;
    if (style instanceof Property) style = style.toStyle(utility.class);
    if (addComment) Array.isArray(style)? style.map((i) => i.property.forEach((p) => (p.comment = className))): style.property.forEach((p) => (p.comment = className));
    return style;
  }
}
