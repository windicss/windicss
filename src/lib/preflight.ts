import { baseUtilities } from './utilities';
import { Style, Property, StyleSheet } from '../utils/style';
import type { ThemeUtil } from '../interfaces';
import type { Processor } from './index';

export default function preflight(
  processor: Processor,
  html?: string,
  includeBase = true,
  includeGlobal = true,
  includePlugins = true,
): StyleSheet {
  // Generate preflight style based on html tags.
  const globalSheet = new StyleSheet();
  const styleSheet = new StyleSheet();
  const pluginSheet = new StyleSheet();

  const createStyle = (
    selector: string | undefined,
    properties: {
      [key: string]: string | string[] | ((theme: ThemeUtil) => string);
    },
    isGlobal = false
  ) => {
    const style = new Style(selector, undefined, false);
    for (const [key, value] of Object.entries(properties)) {
      style.add(
        Array.isArray(value)
          ? value.map((v) => new Property(key, v))
          : new Property(key, typeof value === 'function' ? value((path: string, defaultValue?: unknown) => processor.theme(path, defaultValue)) : value)
      );
    }
    style.updateMeta('base', 'preflight', 0, isGlobal? 1 : 2, true);
    return style;
  };

  const tags = html ? Array.from(new Set(html.match(/<\w+/g))).map((i) => i.substring(1)) : undefined;

  // handle base style
  includeBase && (processor.config('prefixer') ? baseUtilities : baseUtilities.filter(i => !i.selector || !/::?(webkit-input|-moz|-ms-input)-placeholder$/.test(i.selector))).forEach(p => {
    if (includeGlobal && p.global) {
      // global style, such as * or html, body
      globalSheet.add(createStyle(p.selector, p.properties, true));
    } else if (tags !== undefined) {
      // only generate matched styles
      const includeTags = tags.filter((i) => p.keys.includes(i));
      if (includeTags.length > 0) styleSheet.add(createStyle(p.selector ? p.selector : includeTags.join(', '), p.properties));
    } else {
      // if no tags input, generate all styles
      styleSheet.add(createStyle(p.selector ? p.selector : p.keys.join(', '), p.properties));
    }
  });

  // handle plugin style
  if (includePlugins) {
    // base Styles
    let preflightList: Style[] = [];
    Object.values(processor._plugin.preflights).forEach((styles) => {
      preflightList = preflightList.concat(styles);
    });
    pluginSheet.add(preflightList);

    // always generated styles
    let staticList: Style[] = [];
    Object.values(processor._plugin.static).forEach((styles) => {
      staticList = staticList.concat(styles);
    });
    pluginSheet.add(staticList);
  }

  const result = styleSheet.combine().sort().extend(pluginSheet.combine().sort());
  return includeGlobal ? result.extend(globalSheet.combine().sort(), false) : result;
}
