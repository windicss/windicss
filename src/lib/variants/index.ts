import { generateOrientations } from './orientation';
import { generateScreens } from './screen';
import { generateThemes } from './theme';
import { generateStates } from './state';

import type { Style } from '../../utils/style';
import type { BaseTheme, Config, DictStr } from '../../interfaces';

export type Variants = {
  orientation: { [key: string]: () => Style };
  screen: { [key: string]: () => Style };
  theme: { [key: string]: () => Style };
  state: { [key: string]: () => Style };
}

export function resolveVariants(config: Config): Variants {
  return {
    orientation: generateOrientations(((config.theme as BaseTheme)?.orientation ?? {})),
    screen: generateScreens(((config.theme as BaseTheme)?.screens ?? {}) as DictStr),
    theme: generateThemes(config.darkMode),
    state: generateStates(config.variantOrder ?? []),
  };
}

export {
  generateOrientations,
  generateScreens,
  generateThemes,
  generateStates,
};
