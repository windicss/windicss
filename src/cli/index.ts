import arg from 'arg';
import { deepCopy, Console } from '../utils/tools';
import { resolve, dirname, join, extname } from 'path';
import { Processor } from '../lib';
import { readFileSync, writeFile, watch, unwatchFile, existsSync } from 'fs';
import { HTMLParser, CSSParser } from '../utils/parser';
import { StyleSheet } from '../utils/style';
import {
  getVersion,
  globArray,
  generateTemplate,
  fuzzy,
} from './utils';
import type { Extractor } from '../interfaces';

const doc = `Generate css from text files that containing windi classes.
By default, it will use interpretation mode to generate a single css file.

Usage:
  windicss [filenames]
  windicss [filenames] -c -m -d
  windicss [filenames] -c -s -m -d
  windicss [filenames] [-c | -i] [-a] [-b | -s] [-m] [-d] [-p <prefix:string>] [-o <path:string>] [--args arguments]

Options:
  -h, --help            Print this help message and exit.
  -v, --version         Print windicss current version and exit.

  -i, --interpret       Interpretation mode, generate class selectors. This is the default behavior.
  -c, --compile         Compilation mode, combine the class name in each row into a single class.
  -a, --attributify     Attributify mode, generate attribute selectors. Attributify mode can be mixed with the other two modes.
  -t, --preflight       Add preflights, default is false.

  -b, --combine         Combine all css into one single file. This is the default behavior.
  -s, --separate        Generate a separate css file for each input file.

  -d, --dev             Enable hot reload and watch mode.
  -m, --minify          Generate minimized css file.
  -z, --fuzzy           Enable fuzzy match, only works in interpration mode.
  -p, --prefix PREFIX   Set the css class name prefix, only valid in compilation mode. The default prefix is 'windi-'.
  -o, --output PATH     Set output css file path.
  -f, --config PATH     Set config file path.

  --style               Parse and transform windi style block.
  --init PATH           Start a new project on the path.
`;

const args = arg({
  // Types
  '--help': Boolean,
  '--version': Boolean,
  '--compile': Boolean,
  '--interpret': Boolean,
  '--attributify': Boolean,
  '--preflight': Boolean,
  '--combine': Boolean,
  '--separate': Boolean,
  '--dev': Boolean,
  '--minify': Boolean,
  '--fuzzy': Boolean,
  '--style': Boolean,
  '--init': String,
  '--prefix': String,
  '--output': String,
  '--config': String,

  // Aliases
  '-h': '--help',
  '-v': '--version',
  '-i': '--interpret',
  '-c': '--compile',
  '-a': '--attributify',
  '-t': '--preflight',
  '-b': '--combine',
  '-s': '--separate',
  '-d': '--dev',
  '-m': '--minify',
  '-p': '--prefix',
  '-o': '--output',
  '-f': '--config',
  '-z': '--fuzzy',
});

if (args['--help'] || (args._.length === 0 && Object.keys(args).length === 1)) {
  Console.log(doc);
  process.exit();
}

if (args['--version']) {
  Console.log(getVersion());
  process.exit();
}

if (args['--init']) {
  const template = generateTemplate(args['--init'], args['--output']);
  args._.push(template.html);
  args['--preflight'] = true;
  args['--output'] = template.css;
}

const configFile = args['--config'] ? resolve(args['--config']) : undefined;
let preflights: { [key:string]: StyleSheet } = {};
let styleSheets: { [key:string]: StyleSheet } = {};
let processor = new Processor(configFile ? require(configFile) : undefined);
let safelist = processor.config('safelist');

if (configFile) Console.log('Config file:', configFile);

function compile(files: string[]) {
  // compilation mode
  const prefix = args['--prefix'] ?? 'windi-';
  files.forEach((file) => {
    let indexStart = 0;
    const outputStyle: StyleSheet[] = [];
    const outputHTML: string[] = [];
    const html = readFileSync(file).toString();
    const parser = new HTMLParser(html);

    // Match ClassName then replace with new ClassName
    parser.parseClasses().forEach((p) => {
      outputHTML.push(html.substring(indexStart, p.start));
      const utility = processor.compile(p.result, prefix, true, args['--dev']); // Set third argument to false to hide comments;
      outputStyle.push(utility.styleSheet);
      outputHTML.push([utility.className, ...utility.ignored].join(' '));
      indexStart = p.end;
    });
    outputHTML.push(html.substring(indexStart));
    const added = (
      outputStyle.reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
    );
    styleSheets[file] = args['--dev'] ? (styleSheets[file]? styleSheets[file].extend(added) : added) : added;

    const outputFile = file.replace(/(?=\.\w+$)/, '.windi');
    writeFile(outputFile, outputHTML.join(''), () => null);
    Console.log(`${file} -> ${outputFile}`);
    if (args['--preflight']) {
      if (args['--dev']) {
        const preflight = processor.preflight(html, true, true, true, true);
        preflights[file] = preflights[file] ? preflights[file].extend(preflight) : preflight;
      } else {
        preflights[file] = processor.preflight(html);
      }
    }
  });
}

function interpret(files: string[]) {
  // interpretation mode
  files.forEach((file) => {
    const content = readFileSync(file).toString();
    let classes: string[] = [];
    if (args['--fuzzy']) {
      classes = fuzzy(content);
    } else {
      const parser = new HTMLParser(content);
      classes = parser.parseClasses().map((i) => i.result);
    }
    const extractors = processor.config('extract.extractors') as Extractor[] | undefined;
    if (extractors) {
      for (const { extractor, extensions } of extractors) {
        if (extensions.includes(extname(file).slice(1))) {
          const result = extractor(content);
          if ('classes' in result && result.classes) {
            classes = [...classes, ...result.classes];
          }
        }
      }
    }
    if (args['--dev']) {
      const utility = processor.interpret(classes.join(' '), true);
      styleSheets[file] = styleSheets[file] ? styleSheets[file].extend(utility.styleSheet) : utility.styleSheet;
      if (args['--preflight']) {
        const preflight = processor.preflight(content, true, true, true, true);
        preflights[file] = preflights[file] ? preflights[file].extend(preflight) : preflight;
      }
    } else {
      const utility = processor.interpret(classes.join(' '));
      styleSheets[file] = utility.styleSheet;
      if (args['--preflight']) preflights[file] = processor.preflight(content);
    }
  });
}

function attributify(files: string[]) {
  // attributify mode
  files.forEach((file) => {
    const parser = new HTMLParser(readFileSync(file).toString());
    const attrs: { [key: string]: string | string[] } = parser
      .parseAttrs()
      .reduceRight((a: { [key: string]: string | string[] }, b) => {
        if (b.key === 'class' || b.key === 'className') return a;
        if (b.key in a) {
          a[b.key] = Array.isArray(a[b.key])
            ? Array.isArray(b.value)? [ ...(a[b.key] as string[]), ...b.value ]: [ ...(a[b.key] as string[]), b.value ]
            : [ a[b.key] as string, ...(Array.isArray(b.value) ? b.value : [b.value]) ];
          return a;
        }
        return Object.assign(a, { [b.key]: b.value });
      }, {});
    if (args['--dev']) {
      const utility = processor.attributify(attrs, true);
      styleSheets[file] = styleSheets[file] ? styleSheets[file].extend(utility.styleSheet) : utility.styleSheet;
    } else {
      const utility = processor.attributify(attrs);
      styleSheets[file] = utility.styleSheet;
    }
  });
}

function styleBlock(files: string[]) {
  files.forEach((file) => {
    const content = readFileSync(file).toString();
    const block = content.match(/(?<=<style lang=['"]windi["']>)[\s\S]*(?=<\/style>)/);
    if (block && block.index) {
      const css = content.slice(block.index, block.index + block[0].length);
      const parser = new CSSParser(css, processor);
      styleSheets[file] = styleSheets[file].extend(parser.parse());
    }
  });
}

function build(files: string[], update = false) {
  if (args['--compile']) {
    compile(files);
  } else {
    interpret(files);
  }
  if (args['--attributify']) attributify(files);
  if (args['--style']) styleBlock(files);
  if (args['--separate']) {
    for (const [file, sheet] of Object.entries(styleSheets)) {
      const outfile = file.replace(/\.\w+$/, '.windi.css');
      writeFile(outfile, (args['--preflight'] ? deepCopy(sheet).extend(preflights[file], false) : sheet).build(args['--minify']), () => null);
      Console.log(`${file} -> ${outfile}`);
    }
  } else {
    let outputStyle = Object.values(styleSheets)
      .reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
      .sort()
      .combine();
    if (args['--preflight'])
      outputStyle = Object.values(preflights)
        .reduce(
          (previousValue: StyleSheet, currentValue: StyleSheet) =>
            previousValue.extend(currentValue),
          new StyleSheet()
        )
        .sort()
        .combine()
        .extend(outputStyle);
    const filePath = args['--output'] ?? 'windi.css';
    writeFile(filePath, outputStyle.build(args['--minify']), () => null);
    if (!update) {
      Console.log('Matched files:', files);
      Console.log('Output file:', resolve(filePath));
    }
  }

}

function buildSafeList(safelist: unknown) {
  if (safelist) {
    let classes: string[] = [];
    if (typeof safelist === 'string') {
      classes = safelist.split(/\s/).filter(i => i);
    }
    if (Array.isArray(safelist)) {
      for (const item of safelist) {
        if (typeof item === 'string') {
          classes.push(item);
        } else if (Array.isArray(item)) {
          classes = classes.concat(item);
        }
      }
    }
    styleSheets['safelist'] = processor.interpret(classes.join(' ')).styleSheet;
  }
}

const patterns = args._
  .concat(processor.config('extract.include', []) as string[])
  .concat((processor.config('extract.exclude', []) as string[]).map(i => '!' + i));

let matchFiles = globArray(patterns);

if (matchFiles.length === 0) {
  Console.error('No files were matched!');
  process.exit();
}

buildSafeList(safelist);
build(matchFiles);

function watchBuild(file: string) {
  watch(file, (event, path) => {
    if (event === 'rename') {
      const newFiles = globArray(patterns);
      const renamed = matchFiles.filter(i => !(newFiles.includes(i)))[0];
      if (existsSync(path)) {
        Console.log('File', `'${renamed}'`, 'has been renamed to', `'${path}'`);
        matchFiles = newFiles;
        Console.log('Matched files:', matchFiles);
      } else {
        Console.log('File', `'${file}'`, 'has been deleted');
        unwatchFile(file);
        matchFiles = newFiles;
        delete styleSheets[file];
        delete preflights[file];
        if (matchFiles.length > 0) {
          Console.log('Matched files:', matchFiles);
          Console.time('Building');
        }
        build([], true);
        if (matchFiles.length > 0) {
          Console.timeEnd('Building');
        } else {
          Console.error('No files were matched!');
          process.exit();
        }
      }
    }
    if (event === 'change') {
      Console.log('File', `'${path}'`, 'has been changed');
      Console.time('Building');
      build([file], true);
      Console.timeEnd('Building');
    }
  });
}

function watchConfig(file?: string) {
  if (!file) return;
  let stamp = 0;
  watch(file, (event, path) => {
    if (event === 'change' && (stamp === 0 || + new Date() - stamp > 500)) {
      // fix fire twice event when change config file
      stamp = + new Date();
      Console.log('Config', `'${path}'`, 'has been changed');
      Console.time('Building');
      configFile && delete require.cache[configFile];
      processor = new Processor(configFile ? require(configFile) : undefined);
      safelist = processor.config('safelist');
      styleSheets = {};
      preflights = {};
      buildSafeList(safelist);
      build(matchFiles, true);
      Console.timeEnd('Building');
    }
  });
}

if (args['--dev']) {
  watchConfig(configFile);
  for (const file of matchFiles) {
    watchBuild(file);
  }
  for (const dir of Array.from(new Set(matchFiles.map(f => dirname(f))))) {
    watch(dir, (event, path) => {
      if (event === 'rename' && existsSync(join(dir, path))) {
        // when create new file
        const newFiles = globArray(patterns);
        if (newFiles.length > matchFiles.length) {
          const newFile = newFiles.filter(i => !matchFiles.includes(i))[0];
          Console.log('New file', `'${newFile}'`,  'added');
          matchFiles.push(newFile);
          Console.log('Matched files:', matchFiles);
          Console.time('Building');
          build([newFile], true);
          watchBuild(newFile);
          Console.timeEnd('Building');
        }
      }
    });
  }
}
