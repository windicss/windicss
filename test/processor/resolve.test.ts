import { Processor } from '../../src/lib';
import { resolve } from 'path';

const processor = new Processor(require(resolve('./test/assets/windi.plugin.config.js')));

describe('Resolve Tests', () => {
  it('resolve variants', () => {
    const processor = new Processor();
    const screenVariants = [
      'sm',   '<sm',   '@sm', '-sm',  '+sm',
      'md',   '<md',   '@md', '-md',  '+md',
      'lg',   '<lg',   '@lg', '-lg',  '+lg',
      'xl',   '<xl',   '@xl', '-xl',  '+xl',
      '2xl', '<2xl',  '@2xl', '-2xl', '+2xl',
    ];
    const stateVariants = [
      'hover',             'focus',            'active',
      'visited',           'link',             'target',
      'focus-visible',     'focus-within',     'checked',
      'not-checked',       'default',          'disabled',
      'enabled',           'indeterminate',    'invalid',
      'valid',             'optional',         'required',
      'placeholder-shown', 'read-only',        'read-write',
      'not-disabled',      'first-of-type',    'not-first-of-type',
      'last-of-type',      'not-last-of-type', 'first',
      'last',              'not-first',        'not-last',
      'only-child',        'not-only-child',   'only-of-type',
      'not-only-of-type',  'even',             'odd',
      'even-of-type',      'odd-of-type',      'root',
      'empty',             'before',           'after',
      'first-letter',      'first-line',       'file-selector-button',
      'selection',         'svg',              'all',
      'children',          'siblings',         'sibling',
      'ltr',               'rtl',              'group-hover',
      'group-focus',       'group-active',     'group-visited',
      'motion-safe',       'motion-reduce',
    ];
    const themeVariants = [ '@dark', '@light', '.dark', '.light', 'dark', 'light' ];

    expect(Object.keys(processor.resolveVariants())).toEqual([...screenVariants, ...themeVariants, ...stateVariants]);
    expect(Object.keys(processor.resolveVariants('screen'))).toEqual(screenVariants);
    expect(Object.keys(processor.resolveVariants('theme'))).toEqual(themeVariants);
    expect(Object.keys(processor.resolveVariants('state'))).toEqual(stateVariants);
  });

  it('resolve static utilities', () => {
    expect(Object.keys(processor.resolveStaticUtilities(false)).length).toEqual(293);
    expect(Object.keys(processor.resolveStaticUtilities(true)).length).toEqual(299);
  });

  it('resolve dynamic utilities', () => {
    expect(Object.keys(processor.resolveDynamicUtilities(false))).toMatchSnapshot('without-plugins');
    expect(Object.keys(processor.resolveDynamicUtilities(true))).toMatchSnapshot('with-plugins');
  });

  it('get corePlugins', () => {
    const processor = new Processor();
    expect(processor.config('corePlugins')).toMatchSnapshot('list');
    const processor2 = new Processor({
      corePlugins: ['container', 'cursor'],
    });
    expect(processor2.config('corePlugins')).toMatchSnapshot('list2');
    const processor3 = new Processor({
      corePlugins: {
        container: false,
        cursor: false,
      },
    });
    expect(processor3.config('corePlugins')).toMatchSnapshot('list3');
  });
});
