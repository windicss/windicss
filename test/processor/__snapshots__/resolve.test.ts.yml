Tools / get corePlugins / list / 0: |-
  [
    "columns",
    "container",
    "space",
    "divideWidth",
    "divideColor",
    "divideStyle",
    "divideOpacity",
    "accessibility",
    "appearance",
    "backgroundAttachment",
    "backgroundClip",
    "backgroundColor",
    "backgroundImage",
    "gradientColorStops",
    "backgroundOpacity",
    "backgroundPosition",
    "backgroundRepeat",
    "backgroundSize",
    "backgroundOrigin",
    "borderCollapse",
    "borderColor",
    "borderOpacity",
    "borderRadius",
    "borderStyle",
    "borderWidth",
    "boxDecorationBreak",
    "boxSizing",
    "cursor",
    "captionSide",
    "emptyCells",
    "display",
    "flexDirection",
    "flexWrap",
    "placeItems",
    "placeContent",
    "placeSelf",
    "alignItems",
    "alignContent",
    "alignSelf",
    "justifyItems",
    "justifyContent",
    "justifySelf",
    "flex",
    "flexGrow",
    "flexShrink",
    "order",
    "float",
    "clear",
    "fontFamily",
    "fontWeight",
    "height",
    "fontSize",
    "lineHeight",
    "listStylePosition",
    "listStyleType",
    "margin",
    "maxHeight",
    "maxWidth",
    "minHeight",
    "minWidth",
    "objectFit",
    "objectPosition",
    "opacity",
    "outline",
    "overflow",
    "overscrollBehavior",
    "padding",
    "placeholderColor",
    "placeholderOpacity",
    "caretColor",
    "caretOpacity",
    "tabSize",
    "pointerEvents",
    "position",
    "inset",
    "resize",
    "boxShadow",
    "boxShadowColor",
    "ringWidth",
    "ringOffsetColor",
    "ringOffsetWidth",
    "ringColor",
    "ringOpacity",
    "fill",
    "stroke",
    "strokeWidth",
    "strokeDashArray",
    "strokeDashOffset",
    "tableLayout",
    "textAlign",
    "textColor",
    "textOpacity",
    "textOverflow",
    "textShadow",
    "fontStyle",
    "textTransform",
    "textDecorationStyle",
    "textDecorationLength",
    "textDecorationColor",
    "textDecorationOpacity",
    "textDecorationOffset",
    "textDecorationThickness",
    "textDecoration",
    "textIndent",
    "textStrokeColor",
    "textStrokeWidth",
    "content",
    "fontSmoothing",
    "fontVariantNumeric",
    "letterSpacing",
    "userSelect",
    "verticalAlign",
    "visibility",
    "backfaceVisibility",
    "whitespace",
    "wordBreak",
    "writingMode",
    "hyphens",
    "width",
    "zIndex",
    "isolation",
    "gap",
    "gridAutoFlow",
    "gridTemplateColumns",
    "gridAutoColumns",
    "gridColumn",
    "gridColumnStart",
    "gridColumnEnd",
    "gridTemplateRows",
    "gridAutoRows",
    "gridRow",
    "gridRowStart",
    "gridRowEnd",
    "transform",
    "transformOrigin",
    "scale",
    "rotate",
    "translate",
    "skew",
    "perspective",
    "perspectiveOrigin",
    "transitionProperty",
    "transitionTimingFunction",
    "transitionDuration",
    "transitionDelay",
    "keyframes",
    "animation",
    "imageRendering",
    "mixBlendMode",
    "backgroundBlendMode",
    "filter",
    "blur",
    "brightness",
    "contrast",
    "dropShadow",
    "grayscale",
    "hueRotate",
    "invert",
    "saturate",
    "sepia",
    "backdropFilter",
    "backdropBlur",
    "backdropBrightness",
    "backdropContrast",
    "backdropGrayscale",
    "backdropHueRotate",
    "backdropInvert",
    "backdropOpacity",
    "backdropSaturate",
    "backdropSepia",
    "willChange",
    "touchAction",
    "scrollBehavior"
  ]
Tools / get corePlugins / list2 / 1: |-
  [
    "container",
    "cursor"
  ]
Tools / get corePlugins / list3 / 2: |-
  [
    "columns",
    "space",
    "divideWidth",
    "divideColor",
    "divideStyle",
    "divideOpacity",
    "accessibility",
    "appearance",
    "backgroundAttachment",
    "backgroundClip",
    "backgroundColor",
    "backgroundImage",
    "gradientColorStops",
    "backgroundOpacity",
    "backgroundPosition",
    "backgroundRepeat",
    "backgroundSize",
    "backgroundOrigin",
    "borderCollapse",
    "borderColor",
    "borderOpacity",
    "borderRadius",
    "borderStyle",
    "borderWidth",
    "boxDecorationBreak",
    "boxSizing",
    "captionSide",
    "emptyCells",
    "display",
    "flexDirection",
    "flexWrap",
    "placeItems",
    "placeContent",
    "placeSelf",
    "alignItems",
    "alignContent",
    "alignSelf",
    "justifyItems",
    "justifyContent",
    "justifySelf",
    "flex",
    "flexGrow",
    "flexShrink",
    "order",
    "float",
    "clear",
    "fontFamily",
    "fontWeight",
    "height",
    "fontSize",
    "lineHeight",
    "listStylePosition",
    "listStyleType",
    "margin",
    "maxHeight",
    "maxWidth",
    "minHeight",
    "minWidth",
    "objectFit",
    "objectPosition",
    "opacity",
    "outline",
    "overflow",
    "overscrollBehavior",
    "padding",
    "placeholderColor",
    "placeholderOpacity",
    "caretColor",
    "caretOpacity",
    "tabSize",
    "pointerEvents",
    "position",
    "inset",
    "resize",
    "boxShadow",
    "boxShadowColor",
    "ringWidth",
    "ringOffsetColor",
    "ringOffsetWidth",
    "ringColor",
    "ringOpacity",
    "fill",
    "stroke",
    "strokeWidth",
    "strokeDashArray",
    "strokeDashOffset",
    "tableLayout",
    "textAlign",
    "textColor",
    "textOpacity",
    "textOverflow",
    "textShadow",
    "fontStyle",
    "textTransform",
    "textDecorationStyle",
    "textDecorationLength",
    "textDecorationColor",
    "textDecorationOpacity",
    "textDecorationOffset",
    "textDecorationThickness",
    "textDecoration",
    "textIndent",
    "textStrokeColor",
    "textStrokeWidth",
    "content",
    "fontSmoothing",
    "fontVariantNumeric",
    "letterSpacing",
    "userSelect",
    "verticalAlign",
    "visibility",
    "backfaceVisibility",
    "whitespace",
    "wordBreak",
    "writingMode",
    "hyphens",
    "width",
    "zIndex",
    "isolation",
    "gap",
    "gridAutoFlow",
    "gridTemplateColumns",
    "gridAutoColumns",
    "gridColumn",
    "gridColumnStart",
    "gridColumnEnd",
    "gridTemplateRows",
    "gridAutoRows",
    "gridRow",
    "gridRowStart",
    "gridRowEnd",
    "transform",
    "transformOrigin",
    "scale",
    "rotate",
    "translate",
    "skew",
    "perspective",
    "perspectiveOrigin",
    "transitionProperty",
    "transitionTimingFunction",
    "transitionDuration",
    "transitionDelay",
    "keyframes",
    "animation",
    "imageRendering",
    "mixBlendMode",
    "backgroundBlendMode",
    "filter",
    "blur",
    "brightness",
    "contrast",
    "dropShadow",
    "grayscale",
    "hueRotate",
    "invert",
    "saturate",
    "sepia",
    "backdropFilter",
    "backdropBlur",
    "backdropBrightness",
    "backdropContrast",
    "backdropGrayscale",
    "backdropHueRotate",
    "backdropInvert",
    "backdropOpacity",
    "backdropSaturate",
    "backdropSepia",
    "willChange",
    "touchAction",
    "scrollBehavior"
  ]
Tools / resolve dynamic utilities / with-plugins / 1: |-
  [
    "columns",
    "container",
    "space",
    "divide",
    "bg",
    "from",
    "via",
    "to",
    "border",
    "rounded",
    "cursor",
    "flex",
    "order",
    "font",
    "h",
    "leading",
    "list",
    "m",
    "my",
    "mx",
    "mt",
    "mr",
    "mb",
    "ml",
    "min",
    "max",
    "object",
    "opacity",
    "outline",
    "p",
    "py",
    "px",
    "pt",
    "pr",
    "pb",
    "pl",
    "placeholder",
    "caret",
    "tab",
    "indent",
    "inset",
    "top",
    "right",
    "bottom",
    "left",
    "shadow",
    "ring",
    "blur",
    "brightness",
    "contrast",
    "drop",
    "grayscale",
    "hue",
    "invert",
    "saturate",
    "sepia",
    "backdrop",
    "fill",
    "stroke",
    "text",
    "tracking",
    "decoration",
    "underline",
    "w",
    "z",
    "gap",
    "auto",
    "grid",
    "col",
    "row",
    "origin",
    "scale",
    "rotate",
    "translate",
    "skew",
    "perspect",
    "transition",
    "ease",
    "duration",
    "delay",
    "content",
    "animate",
    "filter",
    "aspect-w",
    "aspect-h",
    "aspect",
    "line-clamp",
    "prose"
  ]
Tools / resolve dynamic utilities / without-plugins / 0: |-
  [
    "columns",
    "container",
    "space",
    "divide",
    "bg",
    "from",
    "via",
    "to",
    "border",
    "rounded",
    "cursor",
    "flex",
    "order",
    "font",
    "h",
    "leading",
    "list",
    "m",
    "my",
    "mx",
    "mt",
    "mr",
    "mb",
    "ml",
    "min",
    "max",
    "object",
    "opacity",
    "outline",
    "p",
    "py",
    "px",
    "pt",
    "pr",
    "pb",
    "pl",
    "placeholder",
    "caret",
    "tab",
    "indent",
    "inset",
    "top",
    "right",
    "bottom",
    "left",
    "shadow",
    "ring",
    "blur",
    "brightness",
    "contrast",
    "drop",
    "grayscale",
    "hue",
    "invert",
    "saturate",
    "sepia",
    "backdrop",
    "fill",
    "stroke",
    "text",
    "tracking",
    "decoration",
    "underline",
    "w",
    "z",
    "gap",
    "auto",
    "grid",
    "col",
    "row",
    "origin",
    "scale",
    "rotate",
    "translate",
    "skew",
    "perspect",
    "transition",
    "ease",
    "duration",
    "delay",
    "content",
    "animate"
  ]
